src/System/Random/Stateful.hs

```diff
@@ -0,0 +1,601 @@
+  g <- readSTRef ref
+  case f g of
+    (!a, !g') -> a <$ writeSTRef ref g'
+{-# INLINE applySTGen #-}
+
+-- | Runs a monadic generating action in the `ST` monad using a pure
+-- pseudo-random number generator.
+--
+-- @since 1.2.0
+runSTGen :: RandomGen g => g -> (forall s . STGenM g s -> ST s a) -> (a, g)
+runSTGen g action = unSTGen <$> runST (withMutableGen (STGen g) action)
+
+-- | Runs a monadic generating action in the `ST` monad using a pure
+-- pseudo-random number generator. Returns only the resulting pseudo-random
+-- value.
+--
+-- @since 1.2.0
+runSTGen_ :: RandomGen g => g -> (forall s . STGenM g s -> ST s a) -> a
+runSTGen_ g action = fst $ runSTGen g action
+
+
+-- $uniform
+--
+-- This library provides two type classes to generate pseudo-random values:
+--
+-- *   'UniformRange' is used to generate a value of a type uniformly within a
+--     range.
+-- *   'Uniform' is used to generate a value of a type uniformly over all
+--     possible values of that type.
+--
+-- Types may have instances for both or just one of 'UniformRange' and
+-- 'Uniform'. A few examples illustrate this:
+--
+-- *   'Int', 'Word16' and 'Bool' are instances of both 'UniformRange' and
+--     'Uniform'.
+-- *   'Integer', 'Float' and 'Double' each have an instance for 'UniformRange'
+--     but no 'Uniform' instance.
+-- *   A hypothetical type @Radian@ representing angles by taking values in the
+--     range @[0, 2π)@ has a trivial 'Uniform' instance, but no 'UniformRange'
+--     instance: the problem is that two given @Radian@ values always span /two/
+--     ranges, one clockwise and one anti-clockwise.
+-- *   It is trivial to construct a @Uniform (a, b)@ instance given
+--     @Uniform a@ and @Uniform b@ (and this library provides this tuple
+--     instance).
+-- *   On the other hand, there is no correct way to construct a
+--     @UniformRange (a, b)@ instance based on just @UniformRange a@ and
+--     @UniformRange b@.
+
+-------------------------------------------------------------------------------
+-- Notes
+-------------------------------------------------------------------------------
+
+-- $floating
+--
+-- The 'UniformRange' instances for 'Float' and 'Double' use the following
+-- procedure to generate a random value in a range for @uniformRM (a, b) g@:
+--
+-- 1.  Generate \(x\) uniformly such that \(0 \leq x \leq 1\).
+--
+--     The method by which \(x\) is sampled does not cover all representable
+--     floating point numbers in the unit interval. The method never generates
+--     denormal floating point numbers, for example.
+--
+-- 2.  Return \((b - a) * x + a\).
+--
+--     Due to rounding errors, floating point operations are neither
+--     associative nor distributive the way the corresponding operations on
+--     real numbers are. Additionally, floating point numbers admit special
+--     values @NaN@ as well as negative and positive infinity.
+--
+-- For pathological values, step 2 can yield surprising results.
```

Comment:
With `x * a + (x - 1) * b` plus special cases for `a == b` and infinities, you still get stuff like this:

```
∀ a b x. a < b & 0 < x < 1 ⇒ x × a + (1-x) × b ≤ b
Falsifiable. Counter-example:
  a                   =   -2.13238e-29 :: Float
  b                   = -2.1323799e-29 :: Float
  x                   =     0.27736077 :: Float
  x * a + (x - 1) * b = -2.1323797e-29 :: Float

∀ a b x. a < b & 0 < x < 1 ⇒ a ≤ x × a + (1-x) × b
Falsifiable. Counter-example:
  a                   = -1.9087862 :: Float
  b                   =  -1.908786 :: Float
  x                   =  0.4228573 :: Float
  x * a + (x - 1) * b = -1.9087863 :: Float
```

So there are still _no guarantees_. It does look like the formula you're proposing is slightly more stable, so I'd be happy to make that change. And since `a == b` is very likely to return something that is not equal to either, we can special case that too, provided it's not crazy slow (but that seems unlikely).

---

More results with fewer special cases:

```
∀ a b x. a ≤ b & x = 0 ⇒ a ≤ x × a + (1-x) × b
Falsifiable. Counter-example:
  a                   = -Infinity :: Float
  b                   =  -9.0e-44 :: Float
  x                   =       0.0 :: Float
  x * a + (x - 1) * b =       NaN :: Float

∀ a b x. a ≤ b & x = 1 ⇒ a ≤ x × a + (1-x) × b
Falsifiable. Counter-example:
  a                   = -4.2949693e9 :: Float
  b                   =     Infinity :: Float
  x                   =          1.0 :: Float
  x * a + (x - 1) * b =          NaN :: Float

∀ a b x. a ≤ b & 0 < x < 1 ⇒ a ≤ x × a + (1-x) × b
Falsifiable. Counter-example:
  a                   = -1.2013553e-35 :: Float
  b                   = -1.2013553e-35 :: Float
  x                   =   2.9802322e-8 :: Float
  x * a + (x - 1) * b = -1.2013554e-35 :: Float

∀ a b x. a ≤ b & x = 0 ⇒ x × a + (1-x) × b ≤ b
Falsifiable. Counter-example:
  a                   =     -Infinity :: Float
  b                   = -1.880791e-37 :: Float
  x                   =           0.0 :: Float
  x * a + (x - 1) * b =           NaN :: Float

∀ a b x. a ≤ b & x = 1 ⇒ x × a + (1-x) × b ≤ b
Falsifiable. Counter-example:
  a                   =   -1.125 :: Float
  b                   = Infinity :: Float
  x                   =      1.0 :: Float
  x * a + (x - 1) * b =      NaN :: Float

∀ a b x. a ≤ b & 0 < x < 1 ⇒ x × a + (1-x) × b ≤ b
Falsifiable. Counter-example:
  a                   =  2.2677877 :: Float
  b                   =  2.2677877 :: Float
  x                   = 0.12500384 :: Float
  x * a + (x - 1) * b =   2.267788 :: Float
```

and excluding `a == b` we get:

```
∀ a b x. a < b & x = 0 ⇒ a ≤ x × a + (1-x) × b
Falsifiable. Counter-example:
  a                   = -Infinity :: Float
  b                   =       1.0 :: Float
  x                   =       0.0 :: Float
  x * a + (x - 1) * b =       NaN :: Float

∀ a b x. a < b & x = 1 ⇒ a ≤ x × a + (1-x) × b
Falsifiable. Counter-example:
  a                   =      0.0 :: Float
  b                   = Infinity :: Float
  x                   =      1.0 :: Float
  x * a + (x - 1) * b =      NaN :: Float

∀ a b x. a < b & 0 < x < 1 ⇒ a ≤ x × a + (1-x) × b
Falsifiable. Counter-example:
  a                   =  -2.644899e-38 :: Float
  b                   = -2.6448634e-38 :: Float
  x                   =      0.9998886 :: Float
  x * a + (x - 1) * b = -2.6448993e-38 :: Float

∀ a b x. a < b & x = 0 ⇒ x × a + (1-x) × b ≤ b
Falsifiable. Counter-example:
  a                   = -Infinity :: Float
  b                   =      -0.0 :: Float
  x                   =       0.0 :: Float
  x * a + (x - 1) * b =       NaN :: Float

∀ a b x. a < b & x = 1 ⇒ x × a + (1-x) × b ≤ b
Falsifiable. Counter-example:
  a                   = -2.938379e-39 :: Float
  b                   =      Infinity :: Float
  x                   =           1.0 :: Float
  x * a + (x - 1) * b =           NaN :: Float

∀ a b x. a < b & 0 < x < 1 ⇒ x × a + (1-x) × b ≤ b
Falsifiable. Counter-example:
  a                   = -2.7252144e-38 :: Float
  b                   = -2.5137958e-38 :: Float
  x                   =   4.4703484e-8 :: Float
  x * a + (x - 1) * b = -2.5137955e-38 :: Float
```

Reply:
Whatever is decided on, please leave a note in the source code about the decision

Reply:
See https://github.com/idontgetoutmuch/random/pull/169.